File: src/test_plane.cc

#include <boost/test/unit_test.hpp>
#include "aabb.h"
#include "obb.h"
#include "plane.h"

BOOST_AUTO_TEST_SUITE(plane)

BOOST_AUTO_TEST_CASE(get_origin_1)
{
	rove::plane<> p1(rove::vec<3>(1, -3, 5), rove::vec<3>(-7, 11, -13));
	rove::plane<> p2(p1.get_origin(), p1.get_normal());

	BOOST_REQUIRE(abs(p1.A - p2.A) < rove::EPSILON);
	BOOST_REQUIRE(abs(p1.B - p2.B) < rove::EPSILON);
	BOOST_REQUIRE(abs(p1.C - p2.C) < rove::EPSILON);
	BOOST_REQUIRE(abs(p1.D - p2.D) < rove::EPSILON);
}

BOOST_AUTO_TEST_CASE(classify_point_1)
{
	rove::plane<> pl(rove::vec<3>(0, -1, 0), rove::vec<3>(1, 1, 1));

	BOOST_REQUIRE(pl.classify(rove::vec<3>(0, -2, 0)) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE(pl.classify(rove::vec<3>(0, 0, 0)) == rove::plane<>::POSITIVE);
	BOOST_REQUIRE(pl.classify(rove::vec<3>(0, -1, 0)) == rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_point_2)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(1, -1, 1));

	BOOST_REQUIRE(pl.classify(rove::vec<3>(-2, -1.5, -2)) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE(pl.classify(rove::vec<3>(-1, -2, -1)) == rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_aabb_1)
{
	rove::plane<> pl(rove::vec<3>(-1, -1, -1), rove::vec<3>(1, 1, 1));

	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-2, -2, -2), rove::vec<3>(-1, -1.5, -1))) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-2, -2, -2), rove::vec<3>(-1, -1, -1))) == rove::plane<>::INTERSECTS);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(0, 0, 0), rove::vec<3>(1, 1, 1))) == rove::plane<>::POSITIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-1, -1, -1), rove::vec<3>(1, 1, 1))) == rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_aabb_2)
{
	rove::plane<> pl(rove::vec<3>(-1, -1, -1), rove::vec<3>(-1, -1, -1));

	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-2, -2, -2), rove::vec<3>(-1, -1.5, -1))) == rove::plane<>::POSITIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-2, -2, -2), rove::vec<3>(-1, -1, -1))) == rove::plane<>::INTERSECTS);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(0, 0, 0), rove::vec<3>(1, 1, 1))) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-1, -1, -1), rove::vec<3>(1, 1, 1))) == rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_aabb_3)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(1, -1, 1));

	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(0, 4, 0), rove::vec<3>(1, 5, 1))) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-1, -5, -1), rove::vec<3>(0, -4, 0))) == rove::plane<>::POSITIVE);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-2, -2, -2), rove::vec<3>(0, 0, 0))) == rove::plane<>::INTERSECTS);
	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-1, -1, -1), rove::vec<3>(1, 1, 1))) == rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_aabb_4)
{
	rove::plane<> pl(rove::vec<3>(0, 2, 0), rove::vec<3>(1, 1, 1));

	BOOST_REQUIRE(pl.classify(rove::aabb<3>(rove::vec<3>(-1, -1, -1), rove::vec<3>(0, 0, 0))) == rove::plane<>::NEGATIVE);
}

BOOST_AUTO_TEST_CASE(classify_obb_1)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(1, 1, 1));

	BOOST_REQUIRE(pl.classify(
			rove::obb<3>(rove::vec<3>(0, 1, 0), rove::vec<3>(1, 0, 0), rove::vec<3>(0, 1, 0), rove::vec<3>(0, 0, 1)))
		== rove::plane<>::POSITIVE);

	BOOST_REQUIRE(pl.classify(
			rove::obb<3>(rove::vec<3>(-1, -2, -1), rove::vec<3>(1, 0, 0), rove::vec<3>(0, 1, 0), rove::vec<3>(0, 0, 1)))
		== rove::plane<>::NEGATIVE);
}

BOOST_AUTO_TEST_CASE(classify_obb_2)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(1, 1, 1));

	BOOST_REQUIRE(pl.classify(
			rove::obb<3>(
				rove::vec<3>(-1, -1, -1),
				rove::vec<3>(11, 10, 10),
				rove::vec<3>(10, 11, 10),
				rove::vec<3>(10, 10, 11)))
		== rove::plane<>::INTERSECTS);
}

BOOST_AUTO_TEST_CASE(classify_obb_3)
{
	rove::plane<> pl;
	pl.A =  0.00000000f;
	pl.B = -0.26937401f;
	pl.C = -0.26937401f;
	pl.D =  1.00618340f;

	rove::obb<3> bound;
	bound.origin.set(-5.6245117f, 0.00000000f, 23.716305f);
	bound.tangent.set(2.0000000f, 0.00000000f, 0.00000000f);
	bound.normal.set(0.00000000f, 1.0000000f, 0.00000000f);
	bound.binormal.set(0.00000000f, 0.00000000f, 2.0000000f);

	BOOST_REQUIRE (pl.classify(bound.origin) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE (pl.classify(bound.origin + bound.tangent) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE (pl.classify(bound.origin + bound.normal) == rove::plane<>::NEGATIVE);
	BOOST_REQUIRE (pl.classify(bound.origin + bound.binormal) == rove::plane<>::NEGATIVE);

	BOOST_REQUIRE (pl.classify(bound) == rove::plane<>::NEGATIVE);
}

BOOST_AUTO_TEST_CASE(raytrace_1)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(0, 1, 0));

	rove::ray<3> r(rove::vec<3>(0, 1, 0), rove::vec<3>(0, -10, 0));

	rove::scalar t;
	rove::vec<3> v;

	pl.trace(r, t, v);

	BOOST_REQUIRE((v - rove::vec<3>(0, 0, 0)).length() < rove::EPSILON);
}

BOOST_AUTO_TEST_CASE(raytrace_2)
{
	rove::plane<> pl(rove::vec<3>(0, 0, 0), rove::vec<3>(0, 1, 0));

	rove::ray<3> r(rove::vec<3>(7, 1, 11), rove::vec<3>(0, -10, 0));

	rove::scalar t;
	rove::vec<3> v;

	pl.trace(r, t, v);

	BOOST_REQUIRE((v - rove::vec<3>(7, 0, 11)).length() < rove::EPSILON);
}

BOOST_AUTO_TEST_CASE(raytrace_3)
{
	rove::plane<> pl(rove::vec<3>(1, 3, 5), rove::vec<3>(0, 1, 0));

	rove::ray<3> r(rove::vec<3>(7 + 1, 1 + 3, 11 + 5), rove::vec<3>(0, -10, 0));

	rove::scalar t;
	rove::vec<3> v;

	pl.trace(r, t, v);

	BOOST_REQUIRE((v - rove::vec<3>(7 + 1, 0 + 3, 11 + 5)).length() < rove::EPSILON);
}

BOOST_AUTO_TEST_CASE(raytrace_4)
{
	rove::plane<> pl(rove::vec<3>(1, 3, 5), rove::vec<3>(1, 1, 1));

	rove::ray<3> r(rove::vec<3>(1, 10, 5), rove::vec<3>(0, -10, 0));

	rove::scalar t;
	rove::vec<3> v;

	pl.trace(r, t, v);

	BOOST_REQUIRE((v - rove::vec<3>(1, 3, 5)).length() < rove::EPSILON);
}

BOOST_AUTO_TEST_SUITE_END()

File: src/test_plane.py
import unittest
import sys
import os
import math

# Add build directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'build'))
import pyrove


class TestPlane(unittest.TestCase):
    def test_default_constructor(self):
        # Default constructor leaves values uninitialized (this is expected C++ behavior)
        p = pyrove.plane()
        # Just verify we can create a plane and access properties
        # Values are uninitialized, so we can't test for specific values
        _ = p.A
        _ = p.B
        _ = p.C
        _ = p.D

    def test_constructor_from_origin_normal(self):
        origin = pyrove.vec3(0.0, 0.0, 1.0)
        normal = pyrove.vec3(0.0, 0.0, 1.0)
        p = pyrove.plane(origin, normal)
        # Plane equation: 0*x + 0*y + 1*z + D = 0
        # At origin (0,0,1): 0 + 0 + 1 + D = 0, so D = -1
        self.assertAlmostEqual(p.A, 0.0)
        self.assertAlmostEqual(p.B, 0.0)
        self.assertAlmostEqual(p.C, 1.0)
        self.assertAlmostEqual(p.D, -1.0)

    def test_constructor_from_d_normal(self):
        d = -5.0
        normal = pyrove.vec3(1.0, 0.0, 0.0)
        p = pyrove.plane(d, normal)
        self.assertAlmostEqual(p.A, 1.0)
        self.assertAlmostEqual(p.B, 0.0)
        self.assertAlmostEqual(p.C, 0.0)
        self.assertAlmostEqual(p.D, -5.0)

    def test_set_from_origin_normal(self):
        p = pyrove.plane()
        origin = pyrove.vec3(1.0, 2.0, 3.0)
        normal = pyrove.vec3(0.0, 1.0, 0.0)
        p.set(origin, normal)
        # At point (1,2,3) with normal (0,1,0): 0*x + 1*y + 0*z + D = 0
        # 0 + 2 + 0 + D = 0, so D = -2
        self.assertAlmostEqual(p.A, 0.0)
        self.assertAlmostEqual(p.B, 1.0)
        self.assertAlmostEqual(p.C, 0.0)
        self.assertAlmostEqual(p.D, -2.0)

    def test_set_from_d_normal(self):
        p = pyrove.plane()
        p.set(-3.0, pyrove.vec3(1.0, 1.0, 1.0))
        self.assertAlmostEqual(p.A, 1.0)
        self.assertAlmostEqual(p.B, 1.0)
        self.assertAlmostEqual(p.C, 1.0)
        self.assertAlmostEqual(p.D, -3.0)

    def test_apply(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Points on the plane should evaluate to ~0
        self.assertAlmostEqual(p.apply(pyrove.vec3(1.0, 2.0, 0.0)), 0.0, places=5)

        # Point above the plane
        self.assertAlmostEqual(p.apply(pyrove.vec3(0.0, 0.0, 5.0)), 5.0, places=5)

        # Point below the plane
        self.assertAlmostEqual(p.apply(pyrove.vec3(0.0, 0.0, -3.0)), -3.0, places=5)

    def test_contains_point(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Point on the plane
        self.assertTrue(p.contains(pyrove.vec3(5.0, 3.0, 0.0)))

        # Point not on the plane
        self.assertFalse(p.contains(pyrove.vec3(0.0, 0.0, 0.1)))

    def test_contains_ray(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Ray in the plane
        r1 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(1.0, 0.0, 0.0))
        self.assertTrue(p.contains(r1))

        # Ray not in the plane
        r2 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 1.0), pyrove.vec3(1.0, 0.0, 0.0))
        self.assertFalse(p.contains(r2))

    def test_test_intersection(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Ray that intersects the plane
        r1 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 5.0), pyrove.vec3(0.0, 0.0, -1.0))
        self.assertTrue(p.test_intersection(r1))

        # Ray parallel to the plane
        r2 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 5.0), pyrove.vec3(1.0, 0.0, 0.0))
        self.assertFalse(p.test_intersection(r2))

    def test_parallel(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Ray parallel to the plane
        r1 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 5.0), pyrove.vec3(1.0, 0.0, 0.0))
        self.assertTrue(p.parallel(r1))

        # Ray that intersects the plane
        r2 = pyrove.ray3(pyrove.vec3(0.0, 0.0, 5.0), pyrove.vec3(0.0, 0.0, -1.0))
        self.assertFalse(p.parallel(r2))

    def test_trace(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Ray from (0,0,5) going down should hit plane at t=5
        r = pyrove.ray3(pyrove.vec3(0.0, 0.0, 5.0), pyrove.vec3(0.0, 0.0, -1.0))
        t = p.trace(r)
        self.assertAlmostEqual(t, 5.0, places=5)

    def test_get_normal(self):
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(1.0, 2.0, 3.0))
        n = p.get_normal()
        self.assertAlmostEqual(n.x, 1.0)
        self.assertAlmostEqual(n.y, 2.0)
        self.assertAlmostEqual(n.z, 3.0)

    def test_get_origin(self):
        origin = pyrove.vec3(1.0, 2.0, 3.0)
        normal = pyrove.vec3(0.0, 1.0, 0.0)
        p = pyrove.plane(origin, normal)
        o = p.get_origin()
        # The origin should be on the plane
        self.assertAlmostEqual(p.apply(o), 0.0, places=5)

    def test_is_correct(self):
        # Valid plane
        p1 = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(1.0, 0.0, 0.0))
        self.assertTrue(p1.is_correct())

        # Invalid plane (all coefficients zero)
        p2 = pyrove.plane()
        self.assertFalse(p2.is_correct())

    def test_classify_point(self):
        # XY plane at z=0
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(0.0, 0.0, 1.0))

        # Point above (positive side) - returns 1
        c1 = p.classify(pyrove.vec3(0.0, 0.0, 5.0))
        self.assertEqual(c1, 1)  # POSITIVE

        # Point below (negative side) - returns 3
        c2 = p.classify(pyrove.vec3(0.0, 0.0, -5.0))
        self.assertEqual(c2, 3)  # NEGATIVE

    def test_repr(self):
        p = pyrove.plane(pyrove.vec3(0.0, 0.0, 0.0), pyrove.vec3(1.0, 0.0, 0.0))
        s = repr(p)
        self.assertIn("plane", s)

    def test_property_modification(self):
        p = pyrove.plane()
        p.A = 1.0
        p.B = 2.0
        p.C = 3.0
        p.D = 4.0
        self.assertEqual(p.A, 1.0)
        self.assertEqual(p.B, 2.0)
        self.assertEqual(p.C, 3.0)
        self.assertEqual(p.D, 4.0)


class TestDoublePlane(unittest.TestCase):
    def test_planed(self):
        p = pyrove.planed(pyrove.vec3d(0.0, 0.0, 0.0), pyrove.vec3d(1.0, 0.0, 0.0))
        self.assertAlmostEqual(p.A, 1.0)
        self.assertAlmostEqual(p.B, 0.0)
        self.assertAlmostEqual(p.C, 0.0)

    def test_planed_apply(self):
        p = pyrove.planed(pyrove.vec3d(0.0, 0.0, 0.0), pyrove.vec3d(0.0, 0.0, 1.0))
        result = p.apply(pyrove.vec3d(1.0, 2.0, 5.0))
        self.assertAlmostEqual(result, 5.0)


if __name__ == "__main__":
    unittest.main(verbosity=2)
